platforms/core-runtime/stdlib-java-extensions/src/main/java/org/gradle/api/GradleException.java

```diff
@@ -18,19 +18,43 @@
 
 import org.jspecify.annotations.Nullable;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
 /**
- * <p><code>GradleException</code> is the base class of all exceptions thrown by Gradle.</p>
+ * Intended to be the base class of all exceptions thrown by Gradle.
+ * <p>
+ * Implements {@link ResolutionProvider} to allow the thrower to provide potential resolutions for an exception
+ * that Gradle will display upon failure.
  */
-public class GradleException extends RuntimeException {
+public class GradleException extends RuntimeException implements ResolutionProvider {
+    private final List<String> resolutions = new ArrayList<>();
+
     public GradleException() {
         super();
     }
 
     public GradleException(String message) {
-        super(message);
+        this(message, Collections.emptyList());
+    }
+
+    public GradleException(String message, List<String> resolutions) {
+        this(message, resolutions, null);
     }
 
     public GradleException(String message, @Nullable Throwable cause) {
+        this(message, Collections.emptyList(), cause);
+    }
+
+    public GradleException(String message, List<String> resolutions, @Nullable Throwable cause) {
         super(message, cause);
+        this.resolutions.addAll(resolutions);
+    }
+
+    @Incubating
```

Comment:
What happens if we remove this? It seems weird to put `@Incubating` on the implementation, though if the binary breaking changes plugin requires it then it does...